Frames skipping implementation

Sarah5567 opened the pull request:
**Summary**
Implemented a modular frame comparison component (`FrameChangeDetector`) for selective frame processing in the DeepStream pipeline.  
The logic was integrated into a deepstream pipeline using a pad probe that drops similar frames based on MSE, SSIM, and Optical Flow thresholds.

**Changes**
- Added modular CPU-based frame comparison logic in `frame_change_detector.py`.
- Integrated `FrameChangeDetector` into `deepstream_pipeline.py` via `frame_skip_probe`.
- Verified on sample RTSP video - the pipeline now processes **~21.46%** of frames (skipping similar ones).

> **Note:** currently implemented **on CPU only** -GPU implementation (`cv2.cuda` / `cupy`) pending due to installation issues.

closes #88

Sarah5567 committed:
Implement DeepStream pipeline for frame skipping and processing

Co-authored-by: Sarah Gershuni <sarah556726@gmail.com>
Co-authored-by: Chani Orlinski <c9992946@gmail.com>

Sarah5567 committed:
Refactor FrameChangeDetector to use configurable thresholds and improve preprocessing steps

Co-authored-by: Sarah Gershuni <sarah556726@gmail.com>
Co-authored-by: Chani Orlinski <c9992946@gmail.com>

Sarah5567 committed:
Add opencv and numpy to Dockerfile dependencies

Co-authored-by: Sarah Gershuni <sarah556726@gmail.com>
Co-authored-by: Chani Orlinski <c9992946@gmail.com>

r83575 commented on src/frames_skipping_pipeline.py:
```diff
@@ -0,0 +1,195 @@
+    if not batch_meta:
+        return Gst.PadProbeReturn.OK
+
+    pyds.nvds_acquire_meta_lock(batch_meta)
```
Using `nvds_acquire_meta_lock `despite read-only metadata access.

r83575 commented on src/frames_skipping_pipeline.py:
```diff
@@ -0,0 +1,195 @@
+        except StopIteration:
+            pass
+        finally:
+            pyds.nvds_release_meta_lock(batch_meta)
```
Using `nvds_release_meta_lock` despite read-only metadata access.

Sarah5567 committed:
simplify frame processing logic and ensure proper handling of empty frame lists

Co-authored-by: Sarah Gershuni <sarah556726@gmail.com>
Co-authored-by: Chani Orlinski <c9992946@gmail.com>

lyuzinmaxim commented on env/jetson/Dockerfile:
```diff
@@ -32,7 +32,7 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
 RUN pip3 install --no-cache-dir --upgrade pip wheel setuptools
 
-RUN pip3 install --no-cache-dir isort black flake8 
+RUN pip3 install --no-cache-dir isort black flake8 opencv-python-headless numpy
```
Please merge main - I suppose this is not needed

lyuzinmaxim commented on src/frames_skipping_pipeline.py:
```diff
@@ -0,0 +1,191 @@
+
+detector = FrameChangeDetector()
+
+def should_skip_frame(info, frame_meta, batch_meta):
```
Frame change detector you've initialized (`detector = FrameChangeDetector()`) is not in the context of the function - so once we import the function from another file, it will be gone.

Please check how to make it accessible

Sarah5567 replied:
Do you think it's a proper approach to implement a singleton in `FrameChangeDetector`?

lyuzinmaxim replied:
`frame_skip_probe` runs on each frame kinda indepentently. If you want to access the previous frame, you'd need to have an object which is still "alive" (and which contains the prev frame in our case). Singleton is one of the approaches, yes

lyuzinmaxim commented on src/frames_skipping_pipeline.py:
please run linting and add typings

Sarah5567 committed:
Merge branch 'main' into frames-skipping-Implementation

Sarah5567 committed:
Implement singleton pattern in FrameChangeDetector and integrate it into frame skipping logic

Sarah5567 committed:
apply linting, sorting, and type checking

Co-authored-by: Sarah Gershuni <sarah556726@gmail.com>
Co-authored-by: Chani Orlinski <c9992946@gmail.com>

lyuzinmaxim commented on env/jetson/Dockerfile:
```diff
@@ -115,7 +115,7 @@ RUN mkdir -p /tmp/opencv/build \
 RUN pip3 install --no-cache-dir --upgrade pip wheel setuptools
 
-RUN pip3 install --no-cache-dir isort black flake8 
+RUN pip3 install --no-cache-dir isort black flake8 opencv-python-headless numpy
```
It's not needed. Please see this thread: https://bootcamps2025.slack.com/archives/C09BW2TN7S8/p1762165919662759

lyuzinmaxim commented on src/frame_change_detector.py:
```diff
@@ -33,24 +63,25 @@ def optical_flow_diff(self, imgA: NDArray[np.uint8], imgB: NDArray[np.uint8]) ->
         flow = cv2.calcOpticalFlowFarneback(grayA, grayB, None, 0.5, 3, 15, 3, 5, 1.2, 0)
         mag, _ = cv2.cartToPolar(flow[..., 0], flow[..., 1])
-        return float(np.mean(mag))
+        active = mag[mag > 0.5]
```
I see you added new logic. Please add the "0.5" constant to other contants in this file. Also please add the steps you perform to the existing frame comparison readme

lyuzinmaxim commented on src/frame_change_detector.py:
```diff
@@ -2,14 +2,44 @@
+    _instance = None  # singleton instance
+
+    def __new__(cls, *args, **kwargs):
+        # Create only one instance per process
```
even though here you're checking whether this object was initialized, are you sure that this is actually skipping the frames, i.e. working as expected?

Because here we're initializing new object every frame:
```
detector = FrameChangeDetector()
```

Sarah5567 replied:
Yes, it really skips frames

lyuzinmaxim commented on src/frames_skipping_pipeline.py:
```diff
@@ -0,0 +1,191 @@
+"""
+DeepStream Pipeline
+RTSP → Decode → Convert → Save Images
+"""
```
it's not needed

lyuzinmaxim commented on src/frames_skipping_pipeline.py:
please add typings

Commit:
Fix frame skipping implementation based on code review feedback

- Add OPTICAL_FLOW_ACTIVE_THRESH constant
- Fix singleton pattern with global detector instance
- Remove debug prints
- Clean up code structure

Commit:
Fix frame skipping implementation based on code review feedback

- Add OPTICAL_FLOW_ACTIVE_THRESH constant
- Fix singleton pattern with global detector instance
- Remove debug prints
- Clean up code structure
- Dockerfile already correct (no opencv-python-headless)

Sarah5567 committed:
move FrameChangeDetector initialization into  should_skip_frame

Co-authored-by: Sarah Gershuni <sarah556726@gmail.com>
Co-authored-by: Chani Orlinski <c9992946@gmail.com>

lyuzinmaxim commented on src/frames_skipping_pipeline.py:
```diff
@@ -0,0 +1,177 @@
+os.makedirs(OUTPUT_DIR, exist_ok=True)
+
+
+def should_skip_frame(info: Any, frame_meta: Any, batch_meta: Any) -> int:
```
just as a double-check, could you please create a file called `frame_skipping_callback.py`, and move `should_skip_frame` and `frame_skip_probe` to there, and import those callbacks from it in this file?

Sarah5567 replied:
I did that, still skips frames.

Sarah5567 replied:
This code implements the Singleton pattern because when `FrameChangeDetector()` is called, `__new__` is executed first and checks if `_instance` is `None`. If it is, a new object is created and stored in the class-level variable `_instance`. (I saw a similar implementation [here](https://www.geeksforgeeks.org/python/singleton-pattern-in-python-a-complete-guide/#:~:text=reflected%20in%20samplemodule2.-,Classic%20Singleton%3A,-Classic%20Singleton%20creates)).
After that, `__init__` uses `_initialized` to prevent reinitialization of this single instance.

lyuzinmaxim commented on src/frames_skipping_pipeline.py:
```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python3
```
it's not needed

lyuzinmaxim commented on env/jetson/Dockerfile:
```diff
@@ -115,7 +115,7 @@ RUN mkdir -p /tmp/opencv/build \
 RUN pip3 install --no-cache-dir --upgrade pip wheel setuptools
 
-RUN pip3 install --no-cache-dir isort black flake8 
+RUN pip3 install --no-cache-dir isort black flake8
```
no need to change

Sarah5567 committed:
remove unnecessary shebang

Co-authored-by: Sarah Gershuni <sarah556726@gmail.com>
Co-authored-by: Chani Orlinski <c9992946@gmail.com>

Sarah5567 committed:
restore original whitespace at end of line

Co-authored-by: Sarah Gershuni <sarah556726@gmail.com>
Co-authored-by: Chani Orlinski <c9992946@gmail.com>
